magma/generator.py

```diff
@@ -67,14 +67,22 @@ def _make_type(cls, *args, **kwargs):
     dct = DefineCircuitKind.__prepare__(name, bases)
     cls.__init__(dummy, *args, **kwargs)
     dct.update(dict(dummy.__dict__))
-    return DefineCircuitKind.__new__(cls, name, bases, dct)
+    # Use circuit bind for the instance,
+    # rather than inherited Generator2 bind
+    dct["bind"] = classmethod(DefineCircuitKind.bind)
+    circuit = DefineCircuitKind.__new__(cls, name, bases, dct)
+    for gen in cls.bind_generators:
+        gen.generate_bind(circuit, *args, **kwargs)
+    return circuit
 
 
 class _Generator2Meta(type):
     _cache = weakref.WeakValueDictionary()
 
     def __new__(metacls, name, bases, dct):
         bases = bases + (DefineCircuitKind,)
+        assert "bind_generators" not in dct
+        dct["bind_generators"] = []
```

Comment:
```suggestion
        assert dct.setdefault("bind_generators", []) == []
```

Reply:
Actually, can't just move `bind_generators = []` into the class namespace of `classs Generator2`, i.e.

```python
class Generator2(metaclass=_Generator2Meta):
    bind_generators = []
    ...
```